tracing-core/src/dispatcher.rs

```diff
@@ -682,6 +741,54 @@ impl Drop for DefaultGuard {
     }
 }
 
+#[cfg(not(feature = "std"))]
+impl Drop for DefaultGuard {
+    #[inline]
+    fn drop(&mut self) {
+        if let Some(dispatch) = self.0.take() {
+            if let Some(lock) = LOCK_IMPL {
+                let _guard = lock.lock();
+                let _ = unsafe {
+                    CURRENT_STATE.replace(State {
+                        default: dispatch,
+                        can_enter: true, // TODO(baloo): not too sure
+                    })
+                };
+            }
+        }
+    }
+}
+
+#[cfg(not(feature = "std"))]
+static mut LOCK_IMPL: Option<&'static dyn lock_api::RawMutex<GuardMarker = lock_api::GuardSend>> =
```

Comment:
I don't think this is going to work, since the `RawMutex` trait isn't object safe. I'm not really sure if it's possible for us to be generic over a lock implementation.

Also, the lock we use here will need to be a read-write lock, not a mutex. The current default dispatcher needs to be able to be accessed immutably concurrently; otherwise all spans and events are forced to synchronize the entire program. The write lock is only necessary to _set_ a new default.

This also reduces the danger of deadlocks somewhat. Since recording a trace only requires a read lock, it's okay for interrupt handlers to _record_ a trace, as long as users only call `set_default` when interrupts are disabled.